Graphs/js/held-karp.js

```diff
@@ -0,0 +1,279 @@
+
+    const nums = new Array(n);
+    for (let i = 0; i < n; i++) {
+        nums[i] = i;
+    }
+    nums.splice(x, 1);
+
+    let rst = new Array(n - 1);
+    let min = Infinity;
+    let minArr;
+
+    for (let i = 0; i < nums.length; i++) {
+        let tempNums = nums.concat();
+        tempNums.splice(i, 1);
+        let h = heldKarpPathUndirected(nums[i], x, tempNums, wm, memo);
+        rst[i] = wm[x][nums[i]] + h.distance;
+
+        if (wm[x][nums[i]] + h.distance < min) {
+            min = wm[x][nums[i]] + h.distance;
+            minArr = h.arr.concat();
+        }
+    }
+
+    const minArrInv = minArr.concat();
+
+    minArrInv.reverse().unshift(x);
+
+    let pathList = new LinkedList();
+    let nodes = getAllNodes(cy);
+
+    for (let i = 0; i < minArrInv.length; i++) {
+        pathList.add(nodes[minArrInv[i]]);
+        pathList.add(
+            nodes[minArrInv[i]].edgesWith(nodes[minArrInv[(i + 1) % minArrInv.length]])[0]
+        );
+    }
+
+    clearCyStyle();
+    document.getElementById('ca_weight').innerHTML = min;
+    pathList.traverse(animation_check.checked, true);
+
+    return { distance: min, arr: minArr };
+}
+
+/**
+ *
+ * @param {number} x current node
+ * @param {number} rt starting node
+ * @param {number[]} nums nodes to connect
+ * @param {number[][]} wm distance matrix; wm(ij) represents distance from i to j
+ * @param {Map<string, {distance: number, arr: number[]}>} memo
+ */
+function heldKarpPathUndirected(x, rt, nums, wm, memo) {
+    let nodes = getAllNodes(cy);
+    cy.elements(':grabbable').unselect();
+    if (nums.length === 0) {
+        return { distance: wm[x][rt], arr: [x] };
+    } else {
+        let dist = new Array(nums.length);
+        let min = Infinity;
+        let minArr;
+        for (let i = 0; i < nums.length; i++) {
+            let tempNums = nums.concat();
+            tempNums.splice(i, 1);
+            let key = { from: nums[i], to: x, through: nums };
+            let h;
+            let keyStr = $.map(key, function(val) {
+                return val;
+            }).toString();
+            if (memo.get(keyStr) !== undefined) {
+                h = memo.get(keyStr);
+            } else {
+                h = heldKarpPathUndirected(nums[i], rt, tempNums, wm, memo);
+                memo.set(keyStr, h);
+            }
+
+            dist[i] = wm[x][nums[i]] + h.distance;
+
+            if (dist[i] < min) {
+                min = dist[i];
+                minArr = h.arr.concat();
+                minArr.push(x);
+            }
+        }
+
+        const minArrInv = minArr.concat();
+
+        clearCaStyle();
+        let pathList = new LinkedList();
+        let nodes = getAllNodes(ca);
+
+        for (let i = 0; i < minArrInv.length; i++) {
+            pathList.add(nodes[minArrInv[i]]);
+            pathList.add(
+                nodes[minArrInv[i]].edgesWith(nodes[minArrInv[(i + 1) % minArrInv.length]])[0]
+            );
+            //changed code
```

Comment:
There are not so many changes. Please reduce the code redundancy by merging the undirected and directed functions and add an argument "directed" to the function signature. Also, remember to change the relevant function calls.